File: erpnext_thailand/public/js/payment_entry.js
frappe.ui.form.on("Payment Entry", {

	refresh(frm) {
		// Filter company tax address
		frm.set_query("company_tax_address", function () {
			return {
				filters: {
					is_your_company_address: true,
				},
			};
		});
		// Create Deduct Withholding Tax button
		if (frm.doc.docstatus == 0 && frm.doc.references && frm.doc.references.length > 0) {
			frm.trigger("add_withholding_tax_deduction_buttons");
		}
		// Add button to create withholding tax cert
		if (
			frm.doc.docstatus == 1 &&
			frm.doc.payment_type == "Pay" &&
			frm.doc.deductions.length > 0
		) {
			frm.trigger("add_create_withholding_tax_cert_button");
		}
		// Create Clear Undue VAT Journal Entry
		if (frm.doc.docstatus == 1) {
			frm.trigger("add_create_undue_vat_journal_entry_button");
		}
	},

	add_withholding_tax_deduction_buttons: function (frm) {
		// 1. Based On Manual Selection
		frm.add_custom_button(
			__("Based On Manual Selection"),
			function () {
				frm.trigger("manual_deduct_withholding_tax");
			},
			__("Withhold Tax")
		);
		// 2. Based On Paying Document's Line Items
		frappe.call({
			method: "erpnext_thailand.custom.payment_entry.test_require_withholding_tax",
			args: {
				doc: frm.doc,
			},
			callback: function (r) {
				if (r.message) {
					frm.dashboard.add_comment(
						__(
							"Please be noted that, some paying documents may require <b>Withholding Tax Deduction</b>"
						),
						"blue",
						true
					);
					frm.add_custom_button(
						__("Based On Paying Document's Line Items"),
						function () {
							frm.trigger("auto_deduct_withholding_tax");
						},
						__("Withhold Tax")
					);
				}
			},
		});
	},

	add_create_withholding_tax_cert_button: function (frm) {
		frm.add_custom_button(__("Create Withholding Tax Cert"), async function () {
			let income_tax_form = "";
			if (frm.doc.party_type == "Supplier") {
				supplier_type = (
					await frappe.db.get_value(
						frm.doc.party_type,
						frm.doc.party,
						"supplier_type"
					)
				).message.supplier_type;
				if (supplier_type == "Individual") {
					income_tax_form = "PND3";
				} else {
					income_tax_form = "PND53";
				}
			}
			const fields = [
				{
					fieldtype: "Date",
					label: __("Date"),
					fieldname: "date",
					reqd: 1,
				},
				{
					fieldtype: "Select",
					label: __("Income Tax Form"),
					fieldname: "income_tax_form",
					options: "PND3\nPND53",
					default: income_tax_form,
				},
				{
					fieldtype: "Link",
					label: __("Company Address"),
					fieldname: "company_address",
					options: "Address",
					get_query: () => {
						return {
							filters: {
								is_your_company_address: 1,
							},
						};
					},
				},
			];
			frappe.prompt(
				fields,
				function (filters) {
					frm.events.make_withholding_tax_cert(frm, filters);
				},
				__("Withholding Tax Cert"),
				__("Create Withholding Tax Cert")
			);
		});
	},

	add_create_undue_vat_journal_entry_button: function (frm) {
		// Check first whether all tax has been cleared, to add button
		frappe.call({
			method: "erpnext_thailand.custom.custom_api.to_clear_undue_tax",
			args: {
				dt: cur_frm.doc.doctype,
				dn: cur_frm.doc.name,
			},
			callback: function (r) {
				if (r.message == true) {
					// Add button
					frm.add_custom_button(__("Clear Undue Tax"), function () {
						frm.trigger("make_clear_vat_journal_entry");
					});
				}
			},
		});
	},

	manual_deduct_withholding_tax: function (frm) {
		const fields = [ 
			{
				fieldtype: "Link",
				label: __("WHT Type"),
				fieldname: "wht_type",
				options: "Withholding Tax Type",
				reqd: 1,
				get_query: function () {
					return {
						filters: [[
							"Withholding Tax Type",
							"for_payment_type",
							"in",
							[frm.doc.payment_type, "Pay and Receive"]]],
					};
				},
			},
		];
		frappe.prompt(
			fields,
			(filters) => {
				frm.events.manual_add_withholding_tax_deduction(frm, filters);
			},
			__("Deduct Withholding Tax"),
			__("Add Withholding Tax Deduction")
		);
	},

	auto_deduct_withholding_tax: function (frm) {
		frappe.confirm(
			__(
				"Scan through all reference documents for line items that require tax withholding."
			),
			() => {
				frm.trigger("auto_add_withholding_tax_deduction");
			}
		);
	},

	manual_add_withholding_tax_deduction: function (frm, filters) {
		return frappe.call({
			method: "erpnext_thailand.custom.payment_entry.get_withholding_tax_from_type",
			args: {
				filters: filters,
				doc: frm.doc,
			},
			callback: function (r) {
				var d = frm.add_child("deductions");
				d.withholding_tax_type = r.message["withholding_tax_type"];
				d.account = r.message["account"];
				d.cost_center = r.message["cost_center"];
				d.withholding_tax_base = r.message["base"];
				d.amount = r.message["amount"];
				frm.doc.paid_amount = frm.doc.paid_amount - Math.abs(d.amount);
				frm.refresh();
				frappe.show_alert(
					{
						message: __("Deducted {0} for Withholding Tax", [
							d.amount.toLocaleString(),
						]),
						indicator: "green",
					},
					5
				);
			},
		});
	},

	auto_add_withholding_tax_deduction: function (frm) {
		return frappe.call({
			method: "erpnext_thailand.custom.payment_entry.get_withholding_tax_from_docs_items",
			args: {
				doc: frm.doc,
			},
			callback: function (r) {
				var deduct = 0;
				r.message.forEach(function (item) {
					var d = frm.add_child("deductions");
					d.withholding_tax_type = item["withholding_tax_type"];
					d.account = item["account"];
					d.cost_center = item["cost_center"];
					d.withholding_tax_base = item["base"];
					d.amount = item["amount"];
					deduct += d.amount;
				});
				frm.doc.paid_amount = frm.doc.paid_amount - Math.abs(deduct);
				frm.refresh();
				frappe.show_alert(
					{
						message: __("Deducted {0} for Withholding Tax", [deduct.toLocaleString()]),
						indicator: "green",
					},
					5
				);
			},
		});
	},

	make_withholding_tax_cert: function (frm, filters) {
		return frappe.call({
			method: "erpnext_thailand.custom.payment_entry.make_withholding_tax_cert",
			args: {
				filters: filters,
				doc: frm.doc,
			},
			callback: function (r) {
				var doclist = frappe.model.sync(r.message);
				frappe.set_route("Form", doclist[0].doctype, doclist[0].name);
			},
		});
	},

	make_clear_vat_journal_entry() {
		return frappe.call({
			method: "erpnext_thailand.custom.custom_api.make_clear_vat_journal_entry",
			args: {
				dt: cur_frm.doc.doctype,
				dn: cur_frm.doc.name,
			},
			callback: function (r) {
				var doclist = frappe.model.sync(r.message);
				frappe.set_route("Form", doclist[0].doctype, doclist[0].name);
			},
		});
	},

	// --------------- Thai Billing ---------------

	get_invoices_from_sales_billing: function(frm) {
		const fields = [
			{
                fieldtype: "Link",
                label: __("Sales Billing"),
                fieldname: "sales_billing",
                options: "Sales Billing",
                reqd: 0,
				"get_query": function() {
					return {
						"filters": {
                            "company": frm.doc.company,
                            "customer": frm.doc.party,
                            "docstatus": 1,
                            "total_outstanding_amount": [">", 0]
                        }
					}
				}
			},
			{fieldtype:"Check", label: __("Allocate Payment Amount"), fieldname:"allocate_payment_amount", default:1},
		];

		frappe.prompt(fields, function(filters){
			frm.set_value("sales_billing", filters["sales_billing"]);
			if (!filters["sales_billing"]) { return; }
			frm.events.get_outstanding_documents(frm, filters, true, false);
		}, __("Filters"), __("Get Invoices From Billing"));
	},

    get_invoices_from_purchase_billing: function(frm) {
		const fields = [
			{
                fieldtype: "Link",
                label: __("Purchase Billing"),
                fieldname: "purchase_billing",
                options: "Purchase Billing",
                reqd: 0,
				"get_query": function() {
					return {
						"filters": {
                            "company": frm.doc.company,
                            "supplier": frm.doc.party,
                            "docstatus": 1,
                            "total_outstanding_amount": [">", 0]
                        }
					}
				}
			},
			{fieldtype:"Check", label: __("Allocate Payment Amount"), fieldname:"allocate_payment_amount", default:1},
		];

		frappe.prompt(fields, function(filters){
			frm.set_value("purchase_billing", filters["purchase_billing"]);
			if (!filters["purchase_billing"]) { return; }
			frm.events.get_outstanding_documents(frm, filters, true, false);
		}, __("Filters"), __("Get Invoices From Billing"));
	},

	// --------------- END Thai Billing -----------

});

File: erpnext_thailand/custom/payment_entry.py
import json
from ast import literal_eval

import frappe
import pandas as pd
from frappe import _
from erpnext_thailand.custom.custom_api import get_thai_tax_settings

REF_DOCTYPES = ["Purchase Invoice", "Expense Claim", "Journal Entry"]


@frappe.whitelist()
def test_require_withholding_tax(doc):
	"""Check if any of the payment references has withholding tax type"""
	pay = json.loads(doc)
	for d in pay.get("references"):
		# Purchase Invoice
		if d.get("reference_doctype") in ["Purchase Invoice", "Sales Invoice"]:
			ref_doc = frappe.get_doc(d.get("reference_doctype"), d.get("reference_name"))
			for item in ref_doc.items:
				if item.withholding_tax_type:
					return True
	return False


@frappe.whitelist()
def get_withholding_tax_from_type(filters, doc):
	filters = literal_eval(filters)
	pay = frappe._dict(json.loads(doc))
	wht = frappe.get_doc("Withholding Tax Type", filters["wht_type"])
	company = frappe.get_doc("Company", pay.company)
	base_amount = 0
	for ref in pay.references:
		if ref.get("reference_doctype") not in [
			"Sales Invoice",
			"Purchase Invoice",
			"Expense Claim",
			"Journal Entry",
		]:
			return
		if not ref.get("allocated_amount") or not ref.get("total_amount"):
			continue
		# Find gl entry of ref doc that has undue amount
		gl_entries = frappe.db.get_all(
			"GL Entry",
			filters={
				"voucher_type": ref["reference_doctype"],
				"voucher_no": ref["reference_name"],
			},
			fields=[
				"name",
				"account",
				"debit",
				"credit",
			],
		)
		for gl in gl_entries:
			credit = gl["credit"]
			debit = gl["debit"]
			alloc_percent = ref["allocated_amount"] / ref["total_amount"]
			root_type = frappe.get_cached_value("Account", gl["account"], "root_type")
			account_type = frappe.get_cached_value("Account", gl["account"], "account_type")
			valid_types = (
				"Asset Received But Not Billed",
				"Chargeable",
				"Capital Work in Progress",
				"Cost of Goods Sold",
				"Current Asset",
				"Direct Expense",
				"Direct Income",
				"Expense Account",
				"Expenses Included In Asset Valuation",
				"Expenses Included In Valuation",
				"Fixed Asset",
				"Income Account",
				"Indirect Expense",
				"Indirect Income",
				"Service Received But Not Billed",
				"Temporary"
			)
			if root_type in ["Asset", "Income", "Expense"] and account_type in valid_types:
				base_amount += alloc_percent * (credit - debit)
	if not base_amount:
		frappe.throw(_("There is nothing to withhold tax for"))
	sign = -1 if pay.party_type == "Receive" else 1
	return {
		"withholding_tax_type": wht.name,
		"account": wht.get_account(pay.company),
		"cost_center": company.cost_center,
		"base": base_amount * sign,
		"rate": wht.percent,
		"amount": wht.percent / 100 * base_amount * sign,
	}


@frappe.whitelist()
def get_withholding_tax_from_docs_items(doc):
	pay = frappe._dict(json.loads(doc))
	company = frappe.get_doc("Company", pay.company)
	result = []
	wht_types = frappe.get_all(
		"Withholding Tax Type",
		filters=[
        	["Withholding Tax Type Account", "company", "=", company.name]
        ],
		fields=["name", "percent", "`tabWithholding Tax Type Account`.account"],
		as_list=True,
	)
	wht_rates = frappe._dict({x[0]: {"percent": x[1], "account": x[2]} for x in wht_types})
	for ref in pay.references:
		# For Purchase Invoice, Sales Invoice
		ref_doctype = ref.get("reference_doctype")
		if ref_doctype in ["Purchase Invoice", "Sales Invoice"]:
			if not ref.get("allocated_amount") or not ref.get("total_amount"):
				continue
			sign = -1 if pay.payment_type == "Pay" else 1
			ref_doc = frappe.get_doc(ref_doctype, ref.get("reference_name"))
			for item in ref_doc.items:
				wht_type = get_wht_type(ref_doctype, pay, item)
				if wht_type:
					result.append(
						{
							"withholding_tax_type": wht_type,
							"account": wht_rates[wht_type]["account"],
							"cost_center": company.cost_center,
							"base": sign * item.amount,
							"rate": wht_rates[wht_type]["percent"],
							"amount": wht_rates[wht_type]["percent"] / 100 * sign * item.amount,
						}
					)
	# Group by and sum
	df = pd.DataFrame(result)
	group_fields = ["withholding_tax_type", "account", "cost_center", "rate"]
	sum_fields = ["base", "amount"]
	dict_sum_fields = {x: sum for x in sum_fields}
	result = df.groupby(group_fields, as_index=False).aggregate(dict_sum_fields)
	result = result.to_dict(orient="records")
	return result


def get_wht_type(ref_doctype, pay, item):
	item = frappe.get_doc("Item", item.item_code)
	wht_type = None
	if ref_doctype == "Purchase Invoice":
		supplier = frappe.get_cached_doc("Supplier", pay.party)
		if supplier.supplier_type == "Individual":
			wht_type = item.withholding_tax_type_pay_individual
		else:
			wht_type = item.withholding_tax_type_pay_supplier
	if ref_doctype == "Sales Invoice":
		wht_type = item.withholding_tax_type
	return wht_type


@frappe.whitelist()
def make_withholding_tax_cert(filters, doc):
	filters = literal_eval(filters)
	pay = json.loads(doc)
	cert = frappe.new_doc("Withholding Tax Cert")
	cert.supplier = pay.get("party_type") == "Supplier" and pay.get("party") or ""
	if cert.supplier != "":
		supplier = frappe.get_doc("Supplier", cert.supplier)
		cert.supplier_name = supplier and supplier.supplier_name or ""
		cert.supplier_address = supplier and supplier.supplier_primary_address or ""
	cert.voucher_type = "Payment Entry"
	cert.voucher_no = pay.get("name")
	cert.company_address = filters.get("company_address")
	cert.income_tax_form = filters.get("income_tax_form")
	cert.date = filters.get("date")
	for d in pay.get("deductions"):
		base = d.get("withholding_tax_base", 0)
		amount = d.get("amount", 0)
		rate = 0
		wht_type = d.get("withholding_tax_type")
		if wht_type:
			rate = frappe.get_cached_value("Withholding Tax Type", wht_type, "percent")
		cert.append(
			"withholding_tax_items",
			{
				"tax_base": -base,
				"tax_rate": rate,
				"tax_amount": -amount,
			},
		)
	return cert


def reconcile_undue_tax(doc, method):
	""" If bs_reconcile is installed, unreconcile undue tax gls """
	vouchers = [doc.name] + [r.reference_name for r in doc.references]
	reconcile_undue_tax_gls(vouchers, doc.company)


def reconcile_undue_tax_gls(vouchers, company, unreconcile=False):
	""" Only if bs_reconcile app is install, reconcile/unreconcile undue tax gl entries """
	if "bs_reconcile" not in frappe.get_installed_apps():
		return
	try:
		from bs_reconcile.balance_sheet_reconciliation import utils
	except ImportError:
		pass
	tax = get_thai_tax_settings(company)
	undue_taxes = [tax.purchase_tax_account_undue, tax.sales_tax_account_undue]
	gl_entries = utils.get_gl_entries_by_vouchers(vouchers)
	undue_tax_gls = list(filter(lambda x: x.account in undue_taxes, gl_entries))
	if unreconcile:
		utils.unreconcile_gl(undue_tax_gls)
	else:
		utils.reconcile_gl(undue_tax_gls)


def update_sales_billing_outstanding_amount(doc, method):
	# Document: Payment Entry
	total_outstanding_amount = 0
	if not doc.sales_billing:
		return
	bill = frappe.get_doc("Sales Billing", doc.sales_billing)
	for bill_line in bill.sales_billing_line:
		invoice = frappe.get_doc("Sales Invoice", bill_line.sales_invoice)
		bill_line.outstanding_amount = invoice.outstanding_amount
		total_outstanding_amount += invoice.outstanding_amount
	bill.total_outstanding_amount = total_outstanding_amount
	# Status closed
	bill.closed = 0 if bill.total_outstanding_amount else 1
	bill.save()


@frappe.whitelist()
def get_outstanding_reference_documents(args, validate=False):
	from erpnext.accounts.doctype.payment_entry.payment_entry \
     	import get_outstanding_reference_documents as erpnext_get_outstanding
	data = erpnext_get_outstanding(args, validate)
	# Filter by Sales billing / Purchase Billing
	args = frappe._dict(json.loads(args))
	invoices = []
	if args.sales_billing:
		sales_billing = frappe.get_doc("Sales Billing", args.sales_billing)
		invoices = [x.sales_invoice for x in sales_billing.sales_billing_line]
	elif args.purchase_billing:
		purchase_billing = frappe.get_doc("Purchase Billing", args.purchase_billing)
		invoices = [x.purchase_invoice for x in purchase_billing.purchase_billing_line]
	if invoices:
		data = filter(lambda x: x.get("voucher_no") in invoices, data)
	return data
